src/bank.cpp

```diff
@@ -32,13 +41,60 @@ std::vector<Particle::Bank> master_fission_bank;
 void free_memory_bank()
 {
   simulation::source_bank.clear();
-  #pragma omp parallel
-  {
-    simulation::fission_bank.clear();
+  simulation::fission_bank.reset();
+  simulation::fission_bank_length = 0;
+  simulation::progeny_per_particle.clear();
+}
+
+void init_fission_bank(int64_t max)
+{
+  simulation::fission_bank_max = max;
+  simulation::fission_bank = std::make_unique<Particle::Bank[]>(max);
+  simulation::fission_bank_length = 0;
+  simulation::progeny_per_particle.resize(simulation::work_per_rank);
+}
+
+// Performs an O(n) sort on the fission bank, by leveraging
+// the parent_id and progeny_id fields of banked particles. See the following
+// paper for more details:
+// "Reproducibility and Monte Carlo Eigenvalue Calculations," F.B. Brown and
+// T.M. Sutton, 1992 ANS Annual Meeting, Transactions of the American Nuclear
+// Society, Volume 65, Page 235.
+void sort_fission_bank()
+{
+  // Ensure we don't read off the end of the array if we ran with 0 particles
+  if (simulation::progeny_per_particle.size() == 0) {
+    return;
+  }
+
+  // Perform exclusive scan summation to determine starting indices in fission
+  // bank for each parent particle id
+  int64_t tmp = simulation::progeny_per_particle[0];
+  simulation::progeny_per_particle[0] = 0;
+  for (int64_t i = 1; i < simulation::progeny_per_particle.size(); i++) {
+    int64_t value = simulation::progeny_per_particle[i-1] + tmp;
+    tmp = simulation::progeny_per_particle[i];
+    simulation::progeny_per_particle[i] = value;
+  }
+
+  // TODO: C++17 introduces the exclusive_scan() function which could be
+  // used to replace everything above this point in this function.
+
+  // Create temporary scratch vector for permuting the fission bank
+  std::vector<Particle::Bank> sorted_bank(simulation::fission_bank_length);
```

Comment:
So one thought I had when reading this is that our fission bank is already well over-allocated under normal circumstances. Hypothetically, we could use the upper portion of `fission_bank` to store the sorted sites if the number of banked sites is less than half the size of the fission bank. Not sure how much the memory allocation really matters, but might be worth looking into.

Reply:
That's a good idea! I've added that ability in, so for the typical case it will no longer allocate the extra memory for storing the sorted buffer. It makes the function a little busier, but I think it's worth it to save the memory in most cases.